glib/src/subclass/types.rs

```diff
@@ -233,23 +234,60 @@ impl TypeData {
         self.parent_class
     }
 
-    /// Returns a pointer to the interface implementation specific data.
+    /// Returns a pointer to the class implementation specific data.
     ///
-    /// This is used for interface implementations to store additional data.
-    pub fn get_interface_data(&self, type_: ffi::GType) -> ffi::gpointer {
+    /// This is used for class implementations to store additional data.
+    pub fn get_class_data<T: Any + Send + Sync + 'static>(&self, type_: ffi::GType) -> Option<&T> {
         unsafe {
-            if self.interface_data.is_null() {
-                return ptr::null_mut();
+            if self.class_data.is_none() {
+                return None;
             }
+            let map = self.class_data.unwrap();
+            map.as_ref().get(&type_).map(|v| v.downcast_ref()).flatten()
+        }
+    }
 
-            for &(t, p) in &(*self.interface_data) {
-                if t == type_ {
-                    return p;
-                }
-            }
+    /// Gets a mutable reference of the class implementation specific data
+    ///
+    /// # Safety
+    ///
+    /// This can only be executed before initializing the type data
+    pub unsafe fn get_class_data_mut<T: Any + Send + Sync>(
+        &mut self,
+        type_: ffi::GType,
+    ) -> Option<&mut T> {
+        self.class_data
+            .as_mut()
+            .map(|hashmap| {
+                hashmap
+                    .as_mut()
+                    .get_mut(&type_)
+                    .map(|v| v.downcast_mut())
+                    .flatten()
+            })
+            .flatten()
```

Comment:
Also this `flatten()` can be made an `and_then()` (and you can remove one level of nesting). But the whole thing would probably easier with match and/or `?`